DM-36313: Overhaul registry dataset type and collection wildcards

TallJimbo opened the pull request:
This is a PR based on `tickets/DM-36174` for early review; it will need to be rebased on `main` before merging.

TallJimbo committed:
Overhaul dataset type wildcards and queries.

This adds a new DatasetTypeWildcard class for initial processing of
wildcard expressions and moves the logic for actually resolving those
wildcards into the manager classes.  Dataset type and collection
handling now have roughly analogous code structure.

This also changes the return type of queryDatasetTypes from Iterator
to Iterable, as per RFC-878, and emits deprecation warnings when
patterns or unregistered dataset types are passed to queryDataIds or
queryDimensionRecords, as per RFC-879.

TallJimbo committed:
Overhaul collection wildcards and queries.

This includes:

 - Merging CollectionSearch and CollectionQuery into
   CollectionWildcard, deprecating CollectionSearch (because it was
   publicly visible in a few places) and dropping CollectionQuery.
   "Query" means too many other things, and "wildcard" is much more
   precise as a name, while CollectionSearch is just an immutable
   sequence, and Python has a built-in for that.  And the new class is
   a dataclass, so we don't have to write a bunch of simple
   dunder-methods for it.

 - Moving the code for resolving those wildcards into the
   CollectionsManager implementations, opening the door to
   implementations that don't aggressively fetch all of the
   collections in the entire data repository into memory; I suspect
   we'll want one of those someday.  This actually revealed a bug that
   led to duplicate results, which we were actually (incorrectly!)
   testing _for_ in a CLI test.  That didn't violate the very limited
   guarantees we made about duplicates in query results, but it's
   always nice to get rid of duplicates when it's cheap to do so.

 - Changing the return type of queryCollections to Sequence rather
   than Iterator, since converting the Sequence we start with to an
   Iterator just makes things unnecessarily painful for callers.  This
   was proposed on RFC-878.

TallJimbo committed:
Add collection wildcard resolution interface to QueryBackend.

TallJimbo committed:
Add dataset type wildcard resolution interface to QueryBackend.

TallJimbo committed:
Raise on unregistered dataset types in DCQR.findDatasets.

This implements more of RFC-879.  It allows us to use the new
resolve_single_dataset_type_wildcard method to simplify the
findDatasets method and consistently return a more specific subclass of
DatasetQueryResults.

Some test code was removed here rather than modified because I
discovered that it was totally redundant with other test code that I
did modify.

TallJimbo committed:
Change explain_no_results return type from Iterator to Iterable.

Minor API change from RFC-878, preparing for a future ticket to not
make these lazy iterators at all.

andy-slac commented on python/lsst/daf/butler/registry/wildcards.py:
```diff
@@ -523,120 +407,69 @@ def fromExpression(cls, expression: Any) -> CollectionQuery:
+        -------
+        names : `tuple` [ `str` ]
+            Ordered tuple of collection names.
+        """
```
Maybe add `Raises`?

andy-slac commented on python/lsst/daf/butler/registry/wildcards.py:
```diff
@@ -467,51 +358,44 @@ def __repr__(self) -> str:
+    """
+
+    def __post_init__(self) -> None:
+        if self.patterns is Ellipsis and self.strings:
+            raise ValueError(
+                f"Collection wildcard matches any string, but still has explicit strings {self.strings}."
+            )
```
Is there a chance that we can get `.*` pattern here, which would also match anything?

TallJimbo replied:
We're deprecating (on this branch) *all* patterns in this context, so I don't think it's worth special-casing `.*` temporarily given there there are also lots of almost-as-expansive patterns we can't realistically catch.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_collections.py:
```diff
@@ -585,13 +589,36 @@ def __getitem__(self, key: Any) -> CollectionRecord:
+            If `False`, return records for `~CollectionType.CHAINED`
+            collections themselves.  The default is the opposite of
+            ``flattenChains``: either return records for CHAINED collections or
+            their children, but not both.
```
`done` is not documented.

andy-slac commented on python/lsst/daf/butler/registry/collections/_base.py:
```diff
@@ -456,8 +455,49 @@ def __getitem__(self, key: Any) -> CollectionRecord:
+            for record in self._records.values():
+                if any(p.fullmatch(record.name) for p in wildcard.patterns):
+                    result.extend(resolve_nested(record, done))
+        del resolve_nested
```
Is this needed right before return?

TallJimbo replied:
This is something I've started doing on @natelust 's advice: apparently Python is bad at recognizing when a nested function needs to survive after it goes out of scope - it goes into some sort of delayed garbage-collection channel.  And explicitly deleting the function before return works around that.

andy-slac commented on doc/lsst.daf.butler/queries.rst:
````diff
@@ -30,7 +30,9 @@ Arguments that specify one or more dataset types can generally take any of the f
  - the special value "``...``", which matches all dataset types.
 
-Some of these are not allowed in certain contexts (as documented there).
+Wildcards (`re.Pattern` and ``...```) are not allowed in certain contexts, such as `Registry.queryDataIds` and `Registry.queryDimensionRecords`, particularly when datasets are used only as a constraint what is returned.
````
Triple backticks after ellipsis.
`constraint what is returned` does not quite parse.

andy-slac commented on python/lsst/daf/butler/registry/_registry.py:
```diff
@@ -1218,10 +1218,11 @@ def queryDatasetTypes(
+        Returns
+        -------
+        dataset_types : `Iterable` [ `DatasetType`]
+            An `Iterable `of `DatasetType` instances whose names match
```
Extra space between Iterable and backtick.

andy-slac commented on python/lsst/daf/butler/registry/wildcards.py:
```diff
@@ -473,3 +475,89 @@ def __str__(self) -> str:
+    instances, as the regular constructor performs no checking of inputs (and
+    that can lead to confusing error messages downstream).
+
+    patterns : `tuple` [ `re.Pattern` ]
```
Does this need `Parameters` line?

TallJimbo replied:
Actually I _think_ I should just remove this parameter doc, because this is a dataclass and `__init__` is machine-generated and the docs are on the attributes.  Or at least that's how I usually handle dataclasses, so I hope that's okay.

andy-slac commented on python/lsst/daf/butler/registry/wildcards.py:
```diff
@@ -473,3 +475,89 @@ def __str__(self) -> str:
+            - a `re.Pattern` to match against dataset type names;
+            - an iterable whose elements may be any of the above (any dataset
+              type matching any element in the list is an overall match);
+            - an existing `DatasetTypeWildcard instance;
```
Missing backtick.

andy-slac reviewed (approved):
Looks good, few minor comments.

TallJimbo committed:
Fix bug in wildcard logic.

"coerceUnrecognized" callbacks that return tuples were not previously
supported as documented.

TallJimbo committed:
Add changelog entry for API changes.

andy-slac commented on python/lsst/daf/butler/registry/wildcards.py:
```diff
@@ -90,9 +90,8 @@ def fromExpression(
-            objects of unrecognized type, with the return value added to
-            `strings`. Exceptions will be reraised as `TypeError` (and
-            chained).
+            objects of unrecognized type`. Exceptions will be reraised as
```
Astray backtick.

andy-slac reviewed (approved):
Looks good, one comment.
